### PR TITLE
Stardew Valley: implement new game

### PR DESCRIPTION
**Pull request by Jouramie**
## What is this fixing or adding?
- Add Archipelago implementation for Stardew Valley

## How was this tested?
- 3 play tests with the community, two asyncs and one sync. 
- A lot of feedback and bug reports
- Releases of `.apworld` published https://github.com/agilbert1412/Archipelago/releases

### CODE REVIEW

File: test/TestBase.py
Comment (Berserker66) [inline test/TestBase.py:107]: A bit odd to have these changes in this PR, they do not share the same topic.

Comment (Jouramie) [reply test/TestBase.py:107]: It makes some of the tests easier to understand if we use `self.player` instead of hard coding 1. I can move that in another PR if you want

**Review by black-sliver — COMMENTED** [review]
I started looking at things, here are couple so far:

* we can not really require pytest
  For two reasons:
  1. pycharm has its own thing that (hopefully, some) devs use during their development
  2. pytest's collection code does not pickup packed apworlds and pytest does not honour `load_tests`

  The two "offending" things I see in the PR are
  1. `pytest.mark`
  2. using "bare function" tests (that live outside of a class derived from `unittest.TestCase`)

* Not sure what I think of test.TestBase.WorldTestBase.player

* `worlds/sv/data` has no python code in it, but you have an `__init__.py` there.
  I believe a pure data folder should not be a python package, so the `__init__.py` should probably go.

* A doc string at the top of scripts/*.py would be nice that explains what they are being used for

* I believe frozenset(some_set) will make a copy of the set - so this is kind of wasteful in fish_data.py

* pytest complains about sampling from a set

plus what I commented on inline/below

Oh, and another thing. have not played SV in a while, but shouldn't the shipping container (where you throw stuff in)
be to right of the house?

> Player1 sent Shipping Bin to Player1 (Bamboo Pole Cutscene)

But I don't see a shipping container. Is there another step required after receiving the letter?

**Comment by Jouramie** [comment]
@black-sliver about the __init__.py in data, I believe it's required for the `importlib.resources.path` to work. However, I read yesterday that it's deprecated in latest versions of python, and we should now use `importlib.resources.files`, which is only from 3.9 ... 

Anyway, I work something out for that.

**Comment by Jouramie** [comment]
About pytest, I used it because that's what is used in the workflows. So if it does not pick up apworlds and stuff, the workflows should be checked as well, because the apworlds won't run.

**Comment by Jouramie** [comment]
I'll just remove `test.TestBase.WorldTestBase.player` I guess 🤷

**Comment by agilbert1412** [comment]
> But I don't see a shipping container. Is there another step required after receiving the letter?
All Robin's buildings must be constructed at her shop, after receiving them. But they will be free. This is because they can be placed wherever the player desires, even the shipping bin is not locked to a specific spot

File: worlds/stardew_valley/logic.py
Comment (Jouramie) [reply worlds/stardew_valley/logic.py:95]: It's necessary because
- In the first beta, all the rule was made of lambda, like in other games. However, one cannot really know what's in a lambda without actually running it. It's really now convenient for debugging. 
- Some locations have such complicated logic that it would be extremely hard to reuse and maintain properly. For instances, with the bundle randomization, the rules changes based on the content of the bundle. I'd argue that having a rule for every possible item being already defined really simplified the implementation.  
- In retrospect, the logic framework we built is most likely once of the reason we could release so many features so fast, since it made it really easy to extend the logic.

File: worlds/stardew_valley/options.py
Comment (Jouramie) [reply worlds/stardew_valley/options.py]: We moved the details in [en_Stardew Valley.md](https://github.com/ArchipelagoMW/Archipelago/pull/1455/commits/b1d76e337ab31b735c6d64b27740d838a7997df0#diff-4aa17ab19a8fc4a343976b18bdf7109a06429855c3972012fe9905db79597540)

File: worlds/stardew_valley/options.py
Comment (Jouramie) [reply worlds/stardew_valley/options.py:402]: I removed the None check, however I'd still like to keep the function, since it cast the value in a bool when the option is a Toggle.

**Comment by Jouramie** [comment]
> @black-sliver about the `__init__.py` in data, I believe it's required for the `importlib.resources.path` to work. However, I read yesterday that it's deprecated in latest versions of python, and we should now use `importlib.resources.files`, which is only from 3.9 ...
> 
> Anyway, I work something out for that.

So turns out that it is required with 3.9 to import stuff from the data folder with `importlib.resources`, tho it is not with 3.10. I could find another way if it is really an issue, otherwise I would let it there since it's a really simple fix. I already added some stuff in that file in our dev branch, so it will be coming back next PR anyway.

**Comment by black-sliver** [comment]
> About pytest, I used it because that's what is used in the workflows. So if it does not pick up apworlds and stuff, the workflows should be checked as well, because the apworlds won't run.

The workflow does not have to check packed apworlds, because they are only created when you run `setup.py`, however
  * once we have a world manager, we want to run unit tests on the worlds, so we will have to switch away from pytest
  * people that do not have pytest installed can not import pyest

The solution I would propose (for now) is moving all tests that do not require pytest to classes (that inherit from unittest.TestCase), and tests that require pytest should be skipped if pytest is not installed (via `try: import`)

> So turns out that it is required with 3.9 to import stuff from the data folder with `importlib.resources`

Then it should stay. Sorry I didn't check more closely before writing that.

> All Robin's buildings must be constructed at her shop, after receiving them.

I think it's worth putting that into the tutorial. Everything else that I tested (I only played for a bit) was more or less clear.

File: worlds/stardew_valley/options.py
Comment (black-sliver) [reply worlds/stardew_valley/options.py]: In most cases it is neither required to use `.value` nor is it required to cast it.

E,g, if you have a toggle "do_something", you can do
```
if self.multiworld.do_something[player1]:
    pass
```

You would have to do a lot of `.value` and possibly `bool()` in `def fill_slot_data` though. Maybe it's worth waiting for the generic option dataclass, where we could add a `def as_dict` that auto-casts toggles to bool, not sure.

File: worlds/stardew_valley/logic.py
Comment (black-sliver) [reply worlds/stardew_valley/logic.py:95]: It's definitely an interesting approach. I'm more wondering if it's worth generalizing it, the .difficulty is hard to generalize though.

**Comment by Berserker66** [comment]
I am a bit confused by some statements here. We use pytest for github runs, but otherwise it has always been optional. Personally, on my computer, I run regular unittests through pycharm. Other runners, including pytest, work if so desired.

Making pytest required would be the change, not removing it, as it was never a required dependency to begin with.

**Comment by Jouramie** [comment]
I guess I assumed wrongly that pytest was widely used, since it is used in the CI. My bad on that one. 

All the test have been migrated. All other comments should be addressed as well.

File: worlds/stardew_valley/logic.py
Comment (Jouramie) [reply worlds/stardew_valley/logic.py:95]: It would definitely make logic debugging easier. I'm however not entirely satisfied of the simplification and the difficulty approximation, so that would probably need some improvement before generalizing.

**Review by black-sliver — COMMENTED** [review]
Do we want to have a note that you need to construct all buildings, including the shipping bin, at robin's to the doc? To me it was not clear, so i would recommend documenting that. Having "can now be constructed" it in the letter is probably good enough assuming everybody reads that?

Also is there a limit on client_version yet? I see there are a lot of releases of the mod already, and I have no idea if they what happens if you accidentally use an old mod with a new seed.

After deciding on this and the little detail below, I think this is ready to be merged.

**Comment by agilbert1412** [comment]
> Do we want to have a note that you need to construct all buildings, including the shipping bin, at robin's to the doc? To me it was not clear, so i would recommend documenting that. Having "can now be constructed" it in the letter is probably good enough assuming everybody reads that?

I have added a few lines to guide, explaining in more detail how some items can be redeemed.

> Also is there a limit on client_version yet? I see there are a lot of releases of the mod already, and I have no idea what happens if you accidentally use an old mod with a new seed.

I am actually not sure how we could accomplish that. The mod uses a different versioning scheme than AP itself, so I don't think the standard client_version will work. We cannot submit to AP's versioning scheme because if thinks continue at the same pace, we go through many different versions within one AP release.

I added a comment to the setup page about which version of the mod to use when generating with this instance of the web host. I am open to suggestions on how to enforce it.

As far as I am aware for the mod, things will break very loudly if you use the wrong version, because the slot data will not have information that is considered mandatory. This will not crash the game directly, but between the console being filled with errors and warnings, and the player seeing obvious problems in his world, they probably will not get very far before realizing their mistake

File: worlds/stardew_valley/scripts/export_items.py
Comment (Jouramie) [reply worlds/stardew_valley/scripts/export_items.py]: done

**Comment by black-sliver** [comment]
> I am actually not sure how we could accomplish that. The mod uses a different versioning scheme than AP itself, so I don't think the standard client_version will work. We cannot submit to AP's versioning scheme because if thinks continue at the same pace, we go through many different versions within one AP release.

You can put a custom version thingie in the slot data, set min_client_version to 0.3.9, report 0.3.9 from the mod to ap server, and in addition check if slot data looks compatible to the actual mod, show an error otherwise.

**Review by black-sliver — APPROVED** [review]
Waiting for a reply on the client version thing before hitting merge, but looks good now :-)

**Comment by black-sliver** [comment]
Thanks!

**Pull request merged**
